sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/OptimizeMetadataOnlyDeleteFromTable.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, PredicateHelper, SubqueryExpression}
+import org.apache.spark.sql.catalyst.expressions.Literal.TrueLiteral
+import org.apache.spark.sql.catalyst.plans.logical.{DeleteFromTable, DeleteFromTableWithFilters, LogicalPlan}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.connector.catalog.{SupportsDelete, TruncatableTable}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.sources
+
+/**
+ * Checks whether a delete using filters is possible and nullifies the rewrite plan
+ * if the source can handle this delete without executing the rewrite plan.
+ *
+ * Note this rule must be run after expression optimization but before scan planning.
+ */
+object OptimizeMetadataOnlyDeleteFromTable extends Rule[LogicalPlan] with PredicateHelper {
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case d @ DeleteFromTable(relation: DataSourceV2Relation, cond, Some(_)) =>
```

Comment:
I know in general it's better to put optimizations in optimizer rules, but we can simplify the code quite a lot if we can do this in the rewrite rule `RewriteDeleteFromTable`.

We can statically know if a v2 `Table` extends `SupportsDelete`, if the delete condition contain subquery or not, if the delete condition can be translated to data source filters. which means we don't need to rewrite `DeleteFromTable` when not necessary at the analysis phase.

Then the rewrite rule can be quite simple:
1. If the delete is metadata-only, keep the `DeleteFromTable` unchanged.
2. Otherwise, if the table supports row-level operation, rewrite to `ReplaceData`

This means we don't need to introduce the nested command concept, and we don't need to add a new field to `DeleteFromTable`, and we don't need the new logical plan `DeleteFromTableWithFilters`.

Reply:
I agree it would simplify the implementation but converting the delete condition into data source filters in the analyzer may be suboptimal. Right now, the optimizer rule I added runs after expression optimization, meaning the delete condition has been simplified and rewritten using static optimizer rules.

I followed what we do for `OverwriteByExpression` that also constructs data source filters after expression optimization. That's is also what happens for deletes right now.

To be clear, I am not a fan of adding nested commands or `DeleteFromTableWithFilters`. However, converting the delete condition in the analyzer has substantial limitations as well. Do we have other places that construct data source filters in the analyzer?

Reply:
We can persist the actual DELETE/UPDATE/MERGE condition in `ReplaceData` and use that for job planning but the other concern about converting Catalyst expressions into data source filters without optimizing them remains.

Reply:
I'm thinking about it a bit more. I think it's clearer that we compile group-based DELETE in two steps:
1. create a plan to get affected groups
2. create a plan to get the undeleted rows
These two steps can share one `ScanBuilder` instance.

First, we should have an analyzer rule to rewrite `DeleteFromTable`, similar to what this PR does. But we can make it simpler. We should just replace `DeleteFromTable` with a new plan
```
case class GroupBasedReplaceData(
    table: NamedRelation,
    condition: Expression,
    affectedGroups: LogicalPlan,
    query: LogicalPlan,
    write: Option[Write] = None) extends V2WriteCommand with SupportsSubquery {
}
```
- The first 2 arguments are directly from `DeleteFromTable`, so that we can go back to `DeleteFromTable` later.
- The `affectedGroups` is just a `Filter` with negated DELETE condition, on a DS v2 relation with `RowLevelOperationTable` that is introduced in this PR.
- The `query` is almost the same as `affectedGroups` except that its `Filter` condition is the original DELETE condition.
- The `affectedGroups` and `query` share the same `RowLevelOperationTable` instance, and `RowLevelOperationTable` should create `ScanBuilder` only once, e.g.
```
private lazy val scanBuilder = table.asReadable.newScanBuilder(CaseInsensitiveStringMap.empty())
override def newScanBuilder(options: CaseInsensitiveStringMap): ScanBuilder = scanBuilder
```

Second, we apply a new optimizer rule right before `V2ScanRelationPushDown`. This new rule does predicate pushdown for the `affectedGroups`, to configure `RowLevelOperationTable.scanBuilder`. In the future, when we add APIs to create dedicated v2 `Scan` to get affected groups, this rule can also evaluate the query and collect the affected groups. This rule can also go back to `DeleteFromTable` if we can.

Next, the `V2ScanRelationPushDown` will just work and apply operator pushdown to `GroupBasedReplaceData.query`. Note that, predicate pushdown will be done again here, which is useful: the first predicate pushdown allows the data source to determine which groups to replace, the second predicate pushdown allows the data source to do pruning at runtime (like parquet row group pruning).

Finally, we evaluate `GroupBasedReplaceDataExec`, similar to what this PR does.

Reply:
> Next, the V2ScanRelationPushDown will just work and apply operator pushdown to GroupBasedReplaceData.query. 

I am afraid it is vice versa. The regular planning in `V2ScanRelationPushDown` can be applied to the `filter` subquery, not the `query` plan that contains remaining rows. The `query` plan contains a negated condition that cannot be pushed. That means the main scan must be fully built before `V2ScanRelationPushDown` kicks in. We cannot partially preconfigure the scan there. Parquet pushdown can be enabled in the `filter` subquery, not the main scan.

> this rule can also evaluate the query and collect the affected groups

If we execute the filter subquery in an optimizer rule to form an IN filter while building the main scan, we may not be able to benefit from reusing subqueries and other optimization as it will no longer be one plan. That’s why leveraging the existing framework for runtime filtering seemed so promising.

Based on what we discussed, I removed nested commands by adding more information to `ReplaceData`. I think it simplifies the implementation a little bit. In the future, `ReplaceData` may also contain a filter subquery that will be assigned in the rewrite plan (just like you mentioned). However, let’s put the runtime filtering out of scope for now. We know we have multiple ways to do it: either evaluate a subquery in an optimizer rule or leverage the runtime filtering framework for DS V2. We want to support a custom scan builder anyway. I can work on that immediately after this PR.

What do you think, @cloud-fan?

Reply:
I do have a working prototype for runtime filtering on top of the current implementation that adds the filter subquery to `ReplaceData` and requires an optimizer rule with 50 lines of code. We can use it as a starting point for runtime filtering but I am not sure it should be a blocker to this PR.

Reply:
> I am afraid it is vice versa.

Thanks for correcting! Yea I made a typo and I have fixed my comment now.

> we may not be able to benefit from reusing subqueries and other optimization

You are right. A subquery is better here. The only benefit of evaluating the query in the optimizer I can think of now is that we can get the statistics earlier and work better with CBO. Anyway, we don't need to do it here and we can discuss more in the future.

Reply:
> we can get the statistics earlier and work better with CBO

Yeah, more precise CBO would be nice. I wonder whether adaptive execution can work together with runtime filtering and re-plan after the runtime filter is pushed and we have new and more precise stats reported back by the filtered scan.